README.md
<h1 align="center">Controllerhawk</h1>

<p align="center">The controller/gamepad version of keyhawk, manage your game's keybinds for users using a controller.<p>

<div align="center">

[![NPM version](https://img.shields.io/npm/v/controllerhawk.svg?style=flat)](https://www.npmjs.com/package/controllerhawk)
[![Known Vulnerabilities](https://snyk.io/test/github/robertcorponoi/controllerhawk/badge.svg)](https://snyk.io/test/github/robertcorponoi/controllerhawk)
![npm](https://img.shields.io/npm/dt/controllerhawk)
[![NPM downloads](https://img.shields.io/npm/dm/controllerhawk.svg?style=flat)](https://www.npmjs.com/package/controllerhawk)
<a href="https://badge.fury.io/js/controllerhawk"><img src="https://img.shields.io/github/issues/robertcorponoi/controllerhawk.svg" alt="issues" height="18"></a>
<a href="https://badge.fury.io/js/controllerhawk"><img src="https://img.shields.io/github/license/robertcorponoi/controllerhawk.svg" alt="license" height="18"></a>
[![Gitter](https://badges.gitter.im/gitterHQ/gitter.svg)](https://gitter.im/robertcorponoi)

</div>

Controllerhawk lets you easily and quickly create controller based keybinds for your javascript game. Keybinds are created by assigning one or more buttons to be bound and then passing a function that should be run when the buttons associated with the keybind are used.

## **Installation**

Controllerhawk is an ES6 module that can be used by downloading the package through NPM:

```
$ npm install controllerhawk
```

## **Initialization**

Controllerhawk upon initialization can be passed an options object and for now there is just one option that can be specified.

| param                   | type    | description                                                                                           | default |
|-------------------------|---------|-------------------------------------------------------------------------------------------------------|---------|
| options                 | Object  |                                                                                                       | {}      |
| options.disableGameLoop | boolean | Indicates if Controllerhawk should not use its internal game loop for updating the controller values. | false   |

```js
// Webpack
import Controllerhawk from 'controllerhawk';

// Browser
import Controllerhawk from './path/to/controllerhawk.js';
```

Or you can use it from unpkg like so:

```html
<script type="module" src="https://unpkg.com/controllerhawk@latest/controllerhawk.js">
```

and then lastly you have to initialize it:

```js
// Using the built in game loop:
const controllerhawk = new Controllerhawk();

// Using your own game loop:
const controllerhawk = new Controllerhawk({ disableGameLoop: true });
```

## **Creating Keybinds**

To create a new keybind, you start out by using the `keybind` method which takes a variable amount of arguments depending on how many keys you want to use.

The keys that can be assigned to a keybind should be derived from the `controllerhawk.BUTTONS` object.

**Note:** The when selecting buttons to assign to a keybind, the `instance.BUTTONS` property has three properties of its own, `PS4`, `XBOX` and `GENERIC`. These are more of a convenience so that you can imagine the keybinds better but they will match up with similar buttons on any controller. Here is what the generic gamepad looks like and which all controller layouts will be mapped to:

<img width="500" height="375" src="https://raw.githubusercontent.com/robertcorponoi/graphics/master/controllerhawk/standard-gamepad.png">

Now let's say we have a basic keybind

An example of creating a single key keybind can be done as shown:

```js
// X or cross button on ps4 which ties in to A on xbox or just the botton button on a generic controller.
const jump = controllerhawk.keybind(controllerhawk.BUTTON.PS4.X);

// We could have also defined it as this:
const jump = controllerhawk.keybind(controllerhawk.BUTTON.XBOX.A);

// Lastly with the generic layout:
const jump = controllerhawk.keybind(controllerhawk.BUTTON.GENERIC.QUAD_BUTTON_BOTTOM);
```

The important thing to note here is that all 3 will do the same thing but you only have to define 1. Even if you just define a keybind for a ps4 controller, it will work with the corresponding button on the xbox controller or steam controller or any other standard controller.

You can also add multiple keys by just supplying more parameters like so:

```js
const special = controllerhawk.keybind(controllerhawk.BUTTON.PS4.LEFT_TRIGGER, controllerhawk.BUTTON.PS4.RIGHT_TRIGGER);
```

**Note:** Order of the keys does not matter.

Now a keybind on its own doesn't do much as there's no action associated with it. To add a function to run when the keybind is active, you can specify an action:

```js
const hello = () => {
  console.log('Hello!');
}

const greet = controllerhawk.keybind(controllerhawk.BUTTON.PS4.X).action(hello);
```

This will run the `hello` method everytime the keybind is active.

Notice how `action` can be chained, this is because `keybind` returns an instance of the keybind and so it could be written as:

```js
const hello = () => {
  console.log('Hello!');
}

const greet = controllerhawk.keybind(controllerhawk.BUTTON.PS4.X);

greet.action(hello);
```

## **Using the Analog Sticks**

Most analog sticks on controllers can be pressed so they can also be assigned as buttons to keybinds like with any of the buttons shown above.

However, you probably also want to map movement so you need to know the direction the analog sticks are facing and with how much force. This is where the `leftAnalogStick` and `rightAnalogStick` properties come into play.

These two properties both contain a `x` and `y` property are constantly being updated and the values returned from these will be between -1 and 1. Note that your analog sticks are probably not perfectly aligned and do not sit at 0 and most instead they sit somewhere around -0.01 and 0.01 and it is recommended to check for values beyond that such as `controllerhawk.leftAnalogStick.x > 0.5`.

**x**: The amount that the stick is being pushed in the horizontal direction. A value of -1 will mean the stick is fully to the left while a value of 1 means that the stick is fully to the right and 0 is in the middle.
**y**: The amount that the stick is being pushed in the vertical direction. A value of -1 will mean the stick is fully in the up position while a value of 1 means that the stick is fully in the down position and again 0 is in the middle.

```js
const move = () => {
  sprite.position.x += controllerhawk.leftAnalogStick.x;
};

const jump = () => {
  sprite.position.y += 5 * controllerhawk.rightAnalogStick.y;
};
```

## **Using Your Own Game Loop**

Controllerhawk uses `requestAnimationFrame` interally in order to check for the most up to date values passed from the controller. If you have a game loop running already you may want to bundle the updating that Controllerhawk already does along with it.

In order to do this, you need to set the `disableGameLoop` initialization value to `true` so that Controllerhawk doesn't use the internal loop.

Now to pass it to your own loop, you need to use the `update` method and pass in the current time, which if you're using `requestAnimationFrame` is provided as a parameter.

```js
const controllerhawk = new Controllerhawk({ disableGameLoop: true });

function update(time) {
  controllerhawk.update(time);

  requestAnimationFrame(update);
}

requesetAnimationFrame(update); // Or you could do update(0);
```

## **Examples**

This will be updated with examples using various game engines and templates:

- [Phaser2](#examples/phaser2.md)

## **Tests**

To run the tests for Controllerhawk use:

```bash
npm run test
```

Then in a browser navigate to 'localhost:8888/test/index.html' and you'll be able to see a running demo using a controller.

## **License**

MIT

CHANGELOG.md
## 0.1.1 / 2020-04-16
- [MISC] Updated out-of-date dependencies to their latest versions which also fixed all possible fixed security vulnerabilities.

## 0.1.0 / 2020-02-21
- Initial release.

## 0.0.2 / 2020-02-20
- [FEATURE] Added ability to add keybinds.
- [FEATURE] Added controller mappings for ps4, xbox, and generic controllers.
- [DOCS] Started documenting keybinds.

## 0.0.1 / 2020-02-14
- Initial commit

controllerhawk.js
function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

var Keybind =
/*#__PURE__*/
function () {
  /**
   * The buttons that are assigned to this keybind.
   * 
   * @private
   * 
   * @property {KeybindObject}
   */

  /**
   * The callback method to run when this keybind is used.
   * 
   * @private
   * 
   * @property {Function}
   * 
   * @default this.noop
   */

  /**
   * The timestamp of the last time this keybind was used.
   * 
   * @private
   * 
   * @property {number}
   * 
   * @default 0
   */

  /**
   * @param {KeybindObject} buttons The buttons to bind to this keybind.
   */
  function Keybind(buttons) {
    _classCallCheck(this, Keybind);

    _defineProperty(this, "_buttons", void 0);

    _defineProperty(this, "_action", this._noop);

    _defineProperty(this, "_lastUsed", 0);

    this._buttons = buttons;
  }
  /**
   * Returns the buttons for this keybind.
   * 
   * @returns {Array<number>}
   */


  _createClass(Keybind, [{
    key: "action",

    /**
     * Sets the function to run when this keybind is used.
     * 
     * @param {Function} fn The function to run when this keybind is used.
     * 
     * @returns {Keybind} Returns this for chaining.
     */
    value: function action(fn) {
      this._action = fn;
      return this;
    }
    /**
     * Runs the function associated with this keybind.
     * 
     * @param {number} time The timestamp passed from the game clock.
     */

  }, {
    key: "run",
    value: function run(time) {
      this._action();

      this._lastUsed = time;
    }
    /**
     * An empty function to use as a default value for the action of this keybind.
     * 
     * @private
     */

  }, {
    key: "_noop",
    value: function _noop() {}
  }, {
    key: "buttons",
    get: function get() {
      return this._buttons;
    }
    /**
     * Returns the last time that this keybind was used.
     * 
     * @returns {number}
     */

  }, {
    key: "lastUsed",
    get: function get() {
      return this._lastUsed;
    }
  }]);

  return Keybind;
}();

var Options =
/**
 * Indicates if Controllerhawk should not use its internal game loop for updating the controller values.
 * 
 * @default true
 */

/**
 * @param {Object} options The initialization options passed to Controllerhawk.
 */
function Options(options) {
  _classCallCheck(this, Options);

  _defineProperty(this, "disableGameLoop", false);

  Object.assign(this, options);
};

/**
 * Maps the standard gamepad to a PS4 controller.
 * 
 * https://www.w3.org/TR/gamepad/#remapping
 */

var ps4 = {
  X: 0,
  CROSS: 0,
  CIRCLE: 1,
  SQUARE: 2,
  TRIANGLE: 3,
  LEFT_BUMPER: 4,
  RIGHT_BUMPER: 5,
  LEFT_TRIGGER: 6,
  RIGHT_TRIGGER: 7,
  SELECT: 8,
  START: 9,
  LEFT_ANALOG: 10,
  RIGHT_ANALOG: 11,
  DPAD_UP: 12,
  DPAD_BOTTOM: 13,
  DPAD_LEFT: 14,
  DPAD_RIGHT: 15,
  MIDDLE: 16
};

/**
 * Maps the standard gamepad to a Xbox One controller.
 * 
 * https://www.w3.org/TR/gamepad/#remapping
 */

var xbox = {
  A: 0,
  B: 1,
  X: 2,
  Y: 3,
  LEFT_BUMPER: 4,
  RIGHT_BUMPER: 5,
  LEFT_TRIGGER: 6,
  RIGHT_TRIGGER: 7,
  SELECT: 8,
  START: 9,
  LEFT_ANALOG: 10,
  RIGHT_ANALOG: 11,
  DPAD_UP: 12,
  DPAD_BOTTOM: 13,
  DPAD_LEFT: 14,
  DPAD_RIGHT: 15,
  MIDDLE: 16
};

/**
 * Maps the standard gamepad to a generic controller.
 * 
 * https://www.w3.org/TR/gamepad/#remapping
 */

var generic = {
  QUAD_BUTTON_BOTTOM: 0,
  QUAD_BUTTON_RIGHT: 1,
  QUAD_BUTTON_LEFT: 2,
  QUAD_BUTTON_TOP: 3,
  LEFT_BUMPER: 4,
  RIGHT_BUMPER: 5,
  LEFT_TRIGGER: 6,
  RIGHT_TRIGGER: 7,
  SELECT: 8,
  START: 9,
  LEFT_ANALOG: 10,
  RIGHT_ANALOG: 11,
  DPAD_UP: 12,
  DPAD_BOTTOM: 13,
  DPAD_LEFT: 14,
  DPAD_RIGHT: 15,
  MIDDLE: 16
};

/**
 * Imports all of the individual mappings and provides a way to access any supported controller layout.
 */

var mapping = {
  /**
   * The layout for the ps4 controller.
   */
  PS4: ps4,

  /**
   * The layout for the general xbox/xbox360 controller.
   */
  XBOX: xbox,

  /**
   * The layout for a generic controller.
   */
  GENERIC: generic
};

/**
 * The controller/gamepad version of keyhawk, manage your game's keybinds for users using a controller.
 */
var ControllerHawk =
/*#__PURE__*/
function () {
  /**
   * A reference to the options passed to Controllerhawk.
   * 
   * @private
   * 
   * @property {Options}
   */

  /**
   * A reference to the controllers connected.
   * 
   * @private
   * 
   * @property {Controllers}
   */

  /**
   * A reference to the controllers and their buttons that can be used in keybinds.
   * 
   * @private
   * 
   * @property {Object}
   */

  /**
   * A reference to all of the keybinds that have been created.
   * 
   * @private
   * 
   * @property {Array<Keybind>}
   */

  /**
   * A reference to the buttons on the controller currently being pressed.
   * 
   * @private
   * 
   * @property {Pressed}
   */

  /**
   * The values of the left analog stick.
   * 
   * @private
   * 
   * @property {AnalogStick}
   */

  /**
   * The values of the right analog stick.
   * 
   * @private
   * 
   * @property {AnalogStick}
   */

  /**
   * @param {Object} [options]
   * @param {boolean} [options.disableGameLoop=false] Indicates if Controllerhawk should not use its internal game loop for updating the controller values.
   */
  function ControllerHawk(options) {
    _classCallCheck(this, ControllerHawk);

    _defineProperty(this, "_options", void 0);

    _defineProperty(this, "_controllers", {});

    _defineProperty(this, "_BUTTONS", mapping);

    _defineProperty(this, "_keybinds", []);

    _defineProperty(this, "_pressed", {});

    _defineProperty(this, "_leftAnalogStick", {
      x: 0,
      y: 0
    });

    _defineProperty(this, "_rightAnalogStick", {
      x: 0,
      y: 0
    });

    this._options = new Options(options);

    this._boot();
  }
  /**
   * Returns the controller mappings.
   * 
   * @returns {controller}
   */


  _createClass(ControllerHawk, [{
    key: "keybind",

    /**
     * Creates a new keybind with the specified buttons on a supported controller layout.
     * 
     * @param {...string} buttons One or more buttons from the `CONTROLLER` property to attach to this keybind.
     * 
     * @returns {Keybind} Returns the newly created keybind.
     */
    value: function keybind() {
      for (var _len = arguments.length, buttons = new Array(_len), _key = 0; _key < _len; _key++) {
        buttons[_key] = arguments[_key];
      }

      if (!buttons) {
        console.warn('At least one button must be provided to create a keybind.');
        return;
      }

      var keybind = new Keybind(buttons);

      this._keybinds.push(keybind);

      return keybind;
    }
    /**
     * Checks to see if any keybinds are active and runs them.
     * 
     * @param {number} time The time passed from the game clock.
     */

  }, {
    key: "update",
    value: function update(time) {
      this._update(time);
    }
    /**
     * Adds the necessary event listeners detecting a gamepad connecting and disconnecting.
     * 
     * @private
     */

  }, {
    key: "_boot",
    value: function _boot() {
      var _this = this;

      window.addEventListener('gamepadconnected', function (ev) {
        return _this._onconnect(ev.gamepad);
      });
      window.addEventListener('gamepaddisconnected', function (ev) {
        return _this._ondisconnect(ev.gamepad);
      });
    }
    /**
     * Adds the connected controller to the `_controllers` property and then start the game loop to update the status of 
     * buttons pressed/thumbsticks moved.
     * 
     * @private
     * 
     * @param {Event} ev The `gamepadconnected` event object.
     */

  }, {
    key: "_onconnect",
    value: function _onconnect(gamepad) {
      this._controllers[gamepad.index] = gamepad;
      if (!this._options.disableGameLoop) this._update(0);
    }
    /**
     * Removes the disconnected controller from the `_controllers` property.
     * 
     * @private
     * 
     * @param {Event} ev The `gamepaddisconnected` event object.
     */

  }, {
    key: "_ondisconnect",
    value: function _ondisconnect(gamepad) {
      delete this._controllers[gamepad.index];
    }
    /**
     * Checks to see if there are any gamepads that need to be added and performs controller actions.
     * 
     * @param {number} time The time passed from the game clock.
     * 
     * @private
     */

  }, {
    key: "_update",
    value: function _update(time) {
      var _this2 = this;

      this._scan();

      this._updatePressed();

      this._checkKeybinds(time);

      requestAnimationFrame(function (time) {
        return _this2._update(time);
      });
    }
    /**
     * Checks for active keybinds and runs them.
     * 
     * @param {number} time The time passed from the game clock.
     * 
     * @private
     */

  }, {
    key: "_checkKeybinds",
    value: function _checkKeybinds(time) {
      var _this3 = this;

      this._keybinds.map(function (keybind) {
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = keybind.buttons[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var button = _step.value;
            if (!_this3._pressed[button]) return;
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator["return"] != null) {
              _iterator["return"]();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }

        keybind.run(time);
      });
    }
    /**
     * Updates the status of what buttons are pressed.
     * 
     * @private
     */

  }, {
    key: "_updatePressed",
    value: function _updatePressed() {
      for (var c in this._controllers) {
        var controller = this._controllers[c];

        for (var i = 0; i < controller.buttons.length; ++i) {
          var button = controller.buttons[i];
          if (button.pressed) this._pressed[i] = true;else this._pressed[i] = false;
        }

        var axes = controller.axes;
        this._leftAnalogStick = {
          x: axes[0],
          y: axes[1]
        };
        this._rightAnalogStick = {
          x: axes[2],
          y: axes[3]
        };
      }
    }
    /**
     * Scans for new controllers and if so it adds them to the `_controllers` property or updates the current references.
     * 
     * @private
     */

  }, {
    key: "_scan",
    value: function _scan() {
      var _this4 = this;

      var gamepads = navigator.getGamepads();
      gamepads.map(function (gamepad) {
        if (!gamepad) return;
        if (!(gamepad.index in _this4._controllers)) _this4._onconnect(gamepad);else _this4._controllers[gamepad.index] = gamepad;
      });
    }
  }, {
    key: "BUTTONS",
    get: function get() {
      return this._BUTTONS;
    }
    /**
     * Returns the values of the left analog stick.
     * 
     * @returns {AnalogStick}
     */

  }, {
    key: "leftAnalogStick",
    get: function get() {
      return this._leftAnalogStick;
    }
    /**
     * Returns the values of the right analog stick.
     * 
     * @returns {AnalogStick}
     */

  }, {
    key: "rightAnalogStick",
    get: function get() {
      return this._rightAnalogStick;
    }
  }]);

  return ControllerHawk;
}();

export default ControllerHawk;
